File: trip-scraper/browser/nodejs/trip.mjs
// Trip.com scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// Two surfaces:
// - `scrapeSearch(cityId, checkin, checkout, maxPages)` — `div.hotel-card` items from
//   the city hotel list (`https://www.trip.com/hotels/list?city=<id>`).
// - `scrapeHotel(hotelId, checkin, checkout)` — single property detail page.
//
// Trip.com is less aggressive on anti-bot than Expedia/Priceline, so a single
// goto + selector wait + scroll is usually enough.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 240;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function fetchRenderedHtml(url, readySelector, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 1, scroll = true } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry,
      sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 60000 });
      if (readySelector) {
        try { await page.waitForSelector(readySelector, { timeout: 25000 }); } catch (_) {}
      }
      if (scroll) {
        try {
          await page.evaluate(async () => {
            await new Promise((r) => {
              let y = 0;
              const i = setInterval(() => {
                window.scrollBy(0, 700);
                y += 700;
                if (y >= document.body.scrollHeight) { clearInterval(i); r(); }
              }, 250);
            });
          });
        } catch (_) {}
        await new Promise((r) => setTimeout(r, 2000));
      }
      const html = await page.content();
      if (html) return html;
      lastError = new Error("empty HTML");
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

function parseIntOrNull(text) {
  if (!text) return null;
  const m = /([\d,]+)/.exec(text);
  if (!m) return null;
  const n = parseInt(m[1].replace(/,/g, ""), 10);
  return Number.isFinite(n) ? n : null;
}

function detailUrl(hotelId, checkin, checkout) {
  const params = new URLSearchParams({ hotelId: String(hotelId) });
  if (checkin) params.set("checkIn", checkin);
  if (checkout) params.set("checkOut", checkout);
  return `https://www.trip.com/hotels/detail/?${params.toString()}`;
}

// ---------------- search ----------------

export function parseSearch(html, { checkin = "", checkout = "" } = {}) {
  const $ = cheerio.load(html);
  const out = [];
  // Trip ships two layouts: the older `.hotel-card` and the newer
  // `.compressmeta-hotel-wrap-v8` ("version B"). Match on both via an id-anchored
  // selector with the listing root class.
  $("[id][class*='hotel-card'], [id][class*='compressmeta-hotel-wrap']").each((_, el) => {
    const $c = $(el);
    const id = $c.attr("id") || "";
    if (!/^\d+$/.test(id)) return;
    const name = (
      $c.find(".list-card-title .name").first().text().trim()
      || $c.find(".list-card-title").first().text().trim()
      || $c.find(".hotel-title").first().text().trim()
      || $c.find(".name").first().text().trim()
    );
    // Score: a `.real` block whose text is a plain number (e.g. "9.4").
    let score = null;
    $c.find(".real").each((__, r) => {
      if (score) return;
      const t = $(r).text().trim();
      if (/^\d+(\.\d+)?$/.test(t)) score = t;
    });
    if (!score) score = $c.find(".score").first().text().trim() || null;
    // Word-form review tag near the score.
    const wordEl = $c.find(".describe, .review-rt, .outstanding").first().text();
    const wordMatch = wordEl ? /\b(Outstanding|Excellent|Very Good|Good|Pleasant|Fair|Wonderful|Fabulous|Exceptional)\b/i.exec(wordEl) : null;
    const reviewWord = wordMatch ? wordMatch[1] : null;
    const reviewBlock = $c.find(".count, .review-rt").text();
    const reviewCountMatch = reviewBlock ? /([\d,]+)\s+reviews?/i.exec(reviewBlock) : null;
    const reviewCount = reviewCountMatch ? parseIntOrNull(reviewCountMatch[1]) : null;
    // Price: `.real.labelColor` carries the headline price in version B;
    // older layout uses `.price-line`.
    const price = (
      $c.find(".real.labelColor").first().text().trim()
      || $c.find(".price-line").first().text().trim()
      || null
    );
    const totalPrice = $c.find(".price-explain").first().text().trim() || null;
    const tags = [];
    $c.find(".member-reward-tag, .encourage-tag, .highlight-tag, .hotel-tag").each((__, t) => {
      const txt = $(t).text().trim();
      if (txt && !tags.includes(txt)) tags.push(txt);
    });
    const locText = (
      $c.find(".transport, [class*='location'], [class*='landmark']").first().text().trim().replace(/\s+/g, " ")
      || null
    );
    const image = $c.find(".multi-images img, img.m-lazyImg__img").first().attr("src") || null;
    out.push({
      id,
      name,
      url: detailUrl(id, checkin, checkout),
      score,
      reviewWord,
      reviewCount,
      price,
      totalPrice,
      tags,
      location: locText,
      image,
    });
  });
  return out;
}

export async function scrapeSearch(cityId = "53", checkin = "", checkout = "", maxPages = 1, { proxyCountry = DEFAULT_PROXY_COUNTRY } = {}) {
  const out = [];
  for (let page = 1; page <= maxPages; page++) {
    const params = new URLSearchParams({ city: String(cityId) });
    if (checkin) params.set("checkin", checkin);
    if (checkout) params.set("checkout", checkout);
    if (page > 1) params.set("p", String(page));
    const url = `https://www.trip.com/hotels/list?${params.toString()}`;
    const html = await fetchRenderedHtml(url, "[id][class*='hotel-card'], [id][class*='compressmeta-hotel-wrap']", { proxyCountry });
    const items = parseSearch(html, { checkin, checkout });
    if (!items.length && page > 1) break;
    out.push(...items);
  }
  return out;
}

// ---------------- hotel ----------------

export function parseHotel(html, hotelId, url) {
  const $ = cheerio.load(html);
  // Title can live in `h1.hotelTitle__hotelName`, `.headInfo .name`, or just `h1`.
  const name = (
    $("h1.headInfo .name").first().text().trim()
    || $("h1[class*='hotelName']").first().text().trim()
    || $("h1").first().text().trim()
    || $(".hotel-name").first().text().trim()
  );
  const address = $("[class*='address']").first().text().trim() || null;
  const score = $("[class*='real']").filter((_, el) => /^[0-9.]+$/.test($(el).text().trim())).first().text().trim()
    || $(".score").first().text().trim()
    || null;
  const reviewBlock = $("[class*='comment-num'], [class*='reviewCount']").first().text();
  const reviewCount = parseIntOrNull(reviewBlock);
  const description = (
    $("[class*='introduction']").first().text().trim()
    || $("[class*='hotel-description']").first().text().trim()
    || ""
  );
  const amenities = [];
  $("[class*='facilities'] li, [class*='amenities'] li, [class*='hotelFacility'] li").each((_, el) => {
    const t = $(el).text().trim();
    if (t) amenities.push(t);
  });
  const images = [];
  $("img").each((_, el) => {
    const src = $(el).attr("src") || $(el).attr("data-src");
    if (src && /tripcdn\.com|ak-d\.tripcdn/.test(src) && /hotel|images/i.test(src) && !images.includes(src)) {
      images.push(src);
    }
  });
  const price = $("[class*='price'] [class*='real']").first().text().trim() || null;
  return {
    id: String(hotelId),
    url,
    name,
    address,
    score,
    reviewCount,
    description,
    amenities,
    images: images.slice(0, 30),
    price,
  };
}

export async function scrapeHotel(hotelId, checkin = "", checkout = "", { proxyCountry = DEFAULT_PROXY_COUNTRY } = {}) {
  const url = detailUrl(hotelId, checkin, checkout);
  const html = await fetchRenderedHtml(url, "h1, [class*='headInfo'], [class*='hotelName']", { proxyCountry });
  return parseHotel(html, hotelId, url);
}

File: trip-scraper/browser/python/trip.py
"""Trip.com scraper using the official Scrapeless Python SDK + Playwright over CDP.

Two surfaces:
- ``scrape_search(city_id, checkin, checkout, max_pages)`` — `div.hotel-card` items
  on the Trip.com city hotel list (`https://www.trip.com/hotels/list?city=<id>`).
- ``scrape_hotel(hotel_id, checkin, checkout)`` — single hotel detail page.
"""

from __future__ import annotations

import os
import re
from dataclasses import asdict, dataclass, field
from typing import Any, List, Optional
from urllib.parse import urlencode

from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 240


@dataclass
class SearchResult:
    id: str
    name: str
    url: str
    tags: List[str] = field(default_factory=list)
    score: Optional[str] = None
    reviewWord: Optional[str] = None
    reviewCount: Optional[int] = None
    price: Optional[str] = None
    totalPrice: Optional[str] = None
    location: Optional[str] = None
    image: Optional[str] = None


@dataclass
class Hotel:
    id: str
    url: str
    name: str
    description: str
    amenities: List[str] = field(default_factory=list)
    images: List[str] = field(default_factory=list)
    address: Optional[str] = None
    score: Optional[str] = None
    reviewCount: Optional[int] = None
    price: Optional[str] = None


def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()


async def _fetch_rendered_html(
    url: str,
    ready_selector: Optional[str] = None,
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
    retries: int = 1,
    scroll: bool = True,
) -> str:
    last_error: Exception | None = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
            try:
                page = await browser.new_page()
                await page.goto(url, wait_until="domcontentloaded", timeout=60000)
                if ready_selector:
                    try:
                        await page.wait_for_selector(ready_selector, timeout=25000)
                    except Exception as e:
                        logger.warning("wait_for_selector {!r} failed (continuing): {}", ready_selector, e)
                if scroll:
                    try:
                        await page.evaluate(
                            "() => new Promise(r => { let y = 0; const i = setInterval(() => { window.scrollBy(0, 700); y += 700; if (y >= document.body.scrollHeight) { clearInterval(i); r(); } }, 250); })"
                        )
                    except Exception:
                        pass
                    await page.wait_for_timeout(2000)
                html = await page.content()
                if html:
                    return html
                last_error = RuntimeError("empty HTML")
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")


def _parse_int(text: Optional[str]) -> Optional[int]:
    if not text:
        return None
    m = re.search(r"([\d,]+)", text)
    if not m:
        return None
    try:
        return int(m.group(1).replace(",", ""))
    except ValueError:
        return None


def _detail_url(hotel_id: str, checkin: str = "", checkout: str = "") -> str:
    params: dict[str, str] = {"hotelId": str(hotel_id)}
    if checkin:
        params["checkIn"] = checkin
    if checkout:
        params["checkOut"] = checkout
    return f"https://www.trip.com/hotels/detail/?{urlencode(params)}"


# ---------------- search ----------------


def parse_search(html: str, checkin: str = "", checkout: str = "") -> List[SearchResult]:
    sel = Selector(text=html)
    out: List[SearchResult] = []
    # Trip ships two layouts: the older `.hotel-card` and the newer
    # `.compressmeta-hotel-wrap-v8` ("version B"). Accept both.
    cards = sel.css("[id][class*='hotel-card'], [id][class*='compressmeta-hotel-wrap']")
    for card in cards:
        hid = card.attrib.get("id", "")
        if not hid.isdigit():
            continue
        name = (
            card.css(".list-card-title .name::text").get()
            or card.css(".list-card-title::text").get()
            or card.css(".hotel-title *::text").get()
            or card.css(".hotel-title::text").get()
            or card.css(".name::text").get()
            or ""
        ).strip()
        # Score: a `.real` block whose text is a plain number.
        score: Optional[str] = None
        for t in card.css(".real::text").getall():
            t = t.strip()
            if re.fullmatch(r"\d+(\.\d+)?", t):
                score = t
                break
        if not score:
            score = (card.css(".score::text").get() or "").strip() or None
        word_block = " ".join(card.css(".describe *::text, .review-rt *::text, .outstanding *::text").getall())
        word_m = re.search(r"\b(Outstanding|Excellent|Very Good|Good|Pleasant|Fair|Wonderful|Fabulous|Exceptional)\b", word_block, re.I)
        review_word = word_m.group(1) if word_m else None
        review_block = " ".join(card.css(".count *::text, .review-rt *::text").getall())
        review_count_m = re.search(r"([\d,]+)\s+reviews?", review_block, re.I)
        review_count = _parse_int(review_count_m.group(1)) if review_count_m else None
        price = (
            (card.css(".real.labelColor::text").get() or "").strip()
            or (card.css(".price-line *::text").get() or "").strip()
            or None
        )
        total = "".join(card.css(".price-explain *::text").getall()).strip() or None
        tags: List[str] = []
        for t in card.css(".member-reward-tag *::text, .encourage-tag *::text, .highlight-tag *::text, .hotel-tag *::text").getall():
            t = t.strip()
            if t and t not in tags:
                tags.append(t)
        location_raw = " ".join(card.css(".transport *::text, [class*=location] *::text, [class*=landmark] *::text").getall()).strip()
        location = re.sub(r"\s+", " ", location_raw) or None
        image = card.css(".multi-images img::attr(src), img.m-lazyImg__img::attr(src)").get()
        out.append(
            SearchResult(
                id=hid,
                name=name,
                url=_detail_url(hid, checkin, checkout),
                score=score,
                reviewWord=review_word,
                reviewCount=review_count,
                price=price,
                totalPrice=total,
                tags=tags,
                location=location,
                image=image,
            )
        )
    return out


async def scrape_search(
    city_id: str = "53",
    checkin: str = "",
    checkout: str = "",
    max_pages: int = 1,
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
) -> List[SearchResult]:
    out: List[SearchResult] = []
    for page in range(1, max_pages + 1):
        params: dict[str, str] = {"city": str(city_id)}
        if checkin:
            params["checkin"] = checkin
        if checkout:
            params["checkout"] = checkout
        if page > 1:
            params["p"] = str(page)
        url = f"https://www.trip.com/hotels/list?{urlencode(params)}"
        html = await _fetch_rendered_html(
            url,
            "[id][class*='hotel-card'], [id][class*='compressmeta-hotel-wrap']",
            proxy_country=proxy_country,
        )
        items = parse_search(html, checkin, checkout)
        if not items and page > 1:
            break
        out.extend(items)
    return out


# ---------------- hotel ----------------


def parse_hotel(html: str, hotel_id: str, url: str) -> Hotel:
    sel = Selector(text=html)
    name = (
        sel.css("h1.headInfo .name::text").get()
        or sel.css("h1[class*=hotelName]::text").get()
        or sel.css("h1::text").get()
        or sel.css(".hotel-name::text").get()
        or ""
    ).strip()
    address = (sel.css("[class*=address] *::text").get() or "").strip() or None
    score = (sel.css(".score::text").get() or sel.css("[class*=real]::text").get() or "").strip() or None
    review_block = "".join(sel.css("[class*=comment-num]::text, [class*=reviewCount]::text").getall())
    review_count = _parse_int(review_block)
    description = "".join(sel.css("[class*=introduction] *::text").getall()).strip() \
        or "".join(sel.css("[class*=hotel-description] *::text").getall()).strip()
    amenities = [t.strip() for t in sel.css(
        "[class*=facilities] li *::text, [class*=amenities] li *::text, [class*=hotelFacility] li *::text"
    ).getall() if t.strip()]
    images: List[str] = []
    for src in sel.css("img::attr(src), img::attr(data-src)").getall():
        if not src:
            continue
        if re.search(r"tripcdn\.com|ak-d\.tripcdn", src) and re.search(r"hotel|images", src, re.I):
            if src not in images:
                images.append(src)
    price = (sel.css("[class*=price] [class*=real]::text").get() or "").strip() or None
    return Hotel(
        id=str(hotel_id),
        url=url,
        name=name,
        address=address,
        score=score,
        reviewCount=review_count,
        description=description,
        amenities=amenities,
        images=images[:30],
        price=price,
    )


async def scrape_hotel(
    hotel_id: str,
    checkin: str = "",
    checkout: str = "",
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
) -> Hotel:
    url = _detail_url(hotel_id, checkin, checkout)
    html = await _fetch_rendered_html(
        url,
        ready_selector="h1, [class*=headInfo], [class*=hotelName]",
        proxy_country=proxy_country,
    )
    return parse_hotel(html, hotel_id, url)


def to_dict(obj: Any) -> Any:
    if hasattr(obj, "__dataclass_fields__"):
        return asdict(obj)
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj
